doc/internals.rst

```diff
@@ -42,6 +42,38 @@ xarray objects via the (readonly) :py:attr:`Dataset.variables
 <xarray.Dataset.variables>` and
 :py:attr:`DataArray.variable <xarray.DataArray.variable>` attributes.
 
+Duck arrays
+-----------
+
+.. warning::
+
+    This is a experimental feature.
+
+xarray can wrap custom `duck array`_ objects as long as they define numpy's
+``shape``, ``dtype`` and ``ndim`` properties and the ``__array__``,
+``__array_ufunc__`` and ``__array_function__`` methods.
+
+In certain situations (e.g. when printing the collapsed preview of
+variables of a ``Dataset``), xarray will display the repr of a `duck array`_
+in a single line, truncating it to a certain number of characters. If that
+would drop too much information, the `duck array`_ may define a
+``_repr_inline_`` method that takes ``max_width`` (number of characters) as an
+argument:
```

Comment:
It would be nice to add examples of what should and should not be included in `_repr_inline_`.

In particular, `shape` and `dtype` should *not* be included because xarray already adds those in a standard way.

For examples of what good `_inline_repr_` look like, we could show examples of the output for dask and sparse.